onnxruntime/core/framework/tensor_type_and_shape.cc

```diff
@@ -228,6 +234,7 @@ std::unique_ptr<OrtTensorTypeAndShapeInfo> OrtTensorTypeAndShapeInfo::GetTensorS
 
   if (dim_params != nullptr) {
     type_and_shape->dim_params = *dim_params;
+    type_and_shape->has_shape = true;
```

Comment:
Being able to set `has_shape` to true when `dim_params` is not null seems like a convenient coincidence. Perhaps it would be good to pass that information to this function? I haven't fully thought it through but perhaps something like the following could work:

1. Add a `has_shape` parameter to this `GetTensorShapeAndTypeHelper` function.
2. Callers of `GetTensorShapeAndTypeHelper` can use `onnx::TypeProto_Tensor::has_shape()` to pass this information to `GetTensorShapeAndTypeHelper`.

Reply:
For your suggestion, could you take a look at `onnxruntime/core/framework/onnxruntime_typeinfo.cc:L269`, `L285:L310`. Here already convert the `HasShape` to `nullptr` or a valid pointer. After you review that code, I can add it if you think a new parameter is necessary.